cmd/fixer.go

```diff
@@ -311,6 +316,132 @@ var thumbnailsFixer = &cobra.Command{
 	},
 }
 
+var contactEmailsFixer = &cobra.Command{
+	Use:   "contact-emails",
+	Short: "Detect and try to fix invalid emails on contacts",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		c := newAdminClient()
+		instances, err := c.ListInstances()
+		if err != nil {
+			return err
+		}
+
+		semicolonAtEnd := regexp.MustCompile(";$")
+		dotAtStart := regexp.MustCompile(`^\.`)
+		dotAtEnd := regexp.MustCompile(`\.$`)
+		lessThanStart := regexp.MustCompile("^<")
+		greaterThanEnd := regexp.MustCompile(">$")
+		mailto := regexp.MustCompile("^mailto:")
+
+		fixEmails := func(domain string) error {
+
+			c, err := newClientSafe(domain, consts.Contacts)
+			if err != nil {
+				return err
+			}
+
+			res, err := c.Req(&request.Options{
+				Method: "GET",
+				Path:   "/data/" + consts.Contacts + "/_all_docs",
+				Queries: url.Values{
+					"include_docs": {"true"},
+				},
+			})
+			if err != nil {
+				return err
+			}
+			defer res.Body.Close()
+
+			var contacts struct {
+				Rows []struct {
+					Contact contacts.Contact `json:"doc"`
+				} `json:"rows"`
+			}
+			buf, err := ioutil.ReadAll(res.Body)
+			if err != nil {
+				return err
+			}
+			err = json.Unmarshal(buf, &contacts)
+			if err != nil {
+				return err
+			}
+
+			for _, r := range contacts.Rows {
+				contact := r.Contact
+				id := contact.ID()
+				if strings.HasPrefix(id, "_design") {
+					continue
+				}
+
+				changed := false
+				for _, email := range contact.Email {
+					address := email.Address
+					_, err := mail.ParseAddress(address)
+					if err != nil {
+						old := address
+						address = strings.TrimSpace(address)
+						address = strings.Replace(address, "\"", "", -1)
+						address = strings.Replace(address, ",", ".", -1)
+						address = strings.Replace(address, " .", ".", -1)
+						address = strings.Replace(address, ". ", ".", -1)
+						address = strings.Replace(address, ". ", ".", -1)
+						address = strings.Replace(address, "..", ".", -1)
+						address = strings.Replace(address, ".@", "@", -1)
+						address = strings.Replace(address, "@.", "@", -1)
+						address = strings.Replace(address, " @", "@", -1)
+						address = strings.Replace(address, "@ ", "@", -1)
+						address = mailto.ReplaceAllString(address, "")
+						address = semicolonAtEnd.ReplaceAllString(address, "")
+						address = dotAtStart.ReplaceAllString(address, "")
+						address = dotAtEnd.ReplaceAllString(address, "")
+						address = lessThanStart.ReplaceAllString(address, "")
+						address = greaterThanEnd.ReplaceAllString(address, "")
+						address = strings.TrimSpace(address)
+						_, err := mail.ParseAddress(address)
+						if err == nil {
+							fmt.Printf("    Email fixed: \"%s\" → \"%s\"\n", old, address)
+							changed = true
```

Comment:
I'd have expected something like `changed = address != old`

Reply:
There is a ParseAddress error before and not here, so address was inevitably changed.